### PR TITLE
add OSes to workflow matrix for Python and Rust tests

### LINKED ISSUE #26: Test on Windows
We should test on Windows so we don't miss edge-cases on this platform, for example in path handling.

### PR DESCRIPTION
**Pull request by joshuadavidthomas**
👋 Hiya!

Took a guess at what you meant about testing on Windows in #26 and added an OS key to the workflow matrix for the two test workflows. I went ahead and added macOS as well, but if that's not desired I can always pare it back to Linux and Windows.

closes #26

### CODE REVIEW

**Comment by LilyFoote** [comment]
Hey @joshuadavidthomas, thanks for making a start with this!

To get this PR merged, I think we should get the tests running on each OS, but we don't need all tests passing - we can skip them based on the OS instead and fix them in follow-up PRs. If you feel motivated to fix them in this PR that's good too.

**Comment by joshuadavidthomas** [comment]
@LilyFoote Sounds like a plan! The problems look simple enough (at least at first glance), I don’t mind trying to fix in this PR.

**Comment by joshuadavidthomas** [comment]
Ah damn, you'll have to approve any change for the CI workflows to run. That's going to be a major pain as I try to pick away at this. I'll just do it on my fork for now and bring over the completed fixes all at once.

**Comment by LilyFoote** [comment]
Hmm, I don't appear to be able to approve this latest change. Is there a syntax error or something?

**Comment by joshuadavidthomas** [comment]
> Hmm, I don't appear to be able to approve this latest change. Is there a syntax error or something?

Maybe! I wouldn't worry about it for now -- I've got a PR against my fork I'm working through this on that seems to be working: https://github.com/joshuadavidthomas/django-rusty-templates/pull/7

Only thing that doesn't work is the Codecov action, but minor detail. I've got macOS and Ubuntu done, just hitting my head against Windows at the moment. Right now I've got the different steps intermixed in the single job, but it's getting a bit hard to read (IMO!). Would you mind separate jobs for at least Windows and the Unix based platforms? Or just keep them all together?

**Comment by LilyFoote** [comment]
> > Hmm, I don't appear to be able to approve this latest change. Is there a syntax error or something?
> 
> Maybe! I wouldn't worry about it for now -- I've got a PR against my fork I'm working through this on that seems to be working: [joshuadavidthomas#7](https://github.com/joshuadavidthomas/django-rusty-templates/pull/7)

It only affected one commit, so I'm not worried about it now.

> Only thing that doesn't work is the Codecov action, but minor detail. I've got macOS and Ubuntu done, just hitting my head against Windows at the moment. Right now I've got the different steps intermixed in the single job, but it's getting a bit hard to read (IMO!). Would you mind separate jobs for at least Windows and the Unix based platforms? Or just keep them all together?

Whatever you think is most maintainable is good.

**Comment by joshuadavidthomas** [comment]
@LilyFoote Okay, all greens across the board on CI. Couple of notes:

- Relying on [this repo](https://github.com/vslavik/gettext-tools-windows) to install GNU gettext on Windows, just to get something working for now. It looks relatively active though, with a release from January this year. Can't vouch for it beyond that, so exploring other options later on may be desired.
- Went ahead and activated the virtualenv (via two separate steps, one for Linux/macOS and one for Windows) so the rest of the commands could be the same for all the platforms
- Using bash as the default shell, to get around Windows and calling `source`.
- Changed the `llvm-cov` environment to use a temp shell script to get around shell piping issues on macOS using the `<( ... )` syntax.
- Adjusted one Python test and a handful of Rust tests to account for the differences in path separators on Windows. I went with `#[cfg(not(windows))]` as opposed to `#[cfg(unix)]`, but if you want I can change that.
- Ignored one test on Windows because it looked like a can of worms I didn't necessarily want to open right now. 😄

File: .github/workflows/python.yml [hunk 1]
```diff
@@ -31,29 +56,52 @@ jobs:
...
-        .venv/bin/ruff check
+        ruff check
+
     - name: Test with pytest
       run: |
-        source <(cargo llvm-cov show-env --export-prefix)
-        .venv/bin/pytest --cov --cov-report=xml
+        cargo llvm-cov show-env --export-prefix > llvm-cov-env.sh
+        source llvm-cov-env.sh
+        pytest --cov --cov-report=xml -v
```
Comment (joshuadavidthomas) [inline .github/workflows/python.yml:97]: Ah, the `-v` verbose flag can be removed, this was to see a bit more output from the pytest failure

File: .github/workflows/python.yml [hunk 2]
```diff
@@ -11,18 +11,43 @@ on:
...
+      if: runner.os == 'Linux'
       run: sudo apt-get install gettext
 
+    - name: Install gettext for translation testing (macOS)
+      if: runner.os == 'macOS'
+      run: brew install gettext
+
+    - name: Install gettext for translation testing (Windows)
+      if: runner.os == 'Windows'
+      shell: bash
```
Comment (joshuadavidthomas) [inline .github/workflows/python.yml:43]: default is bash now, this can be removed

**Comment by LilyFoote** [comment]
>     * Relying on [this repo](https://github.com/vslavik/gettext-tools-windows) to install GNU gettext on Windows, just to get something working for now. It looks relatively active though, with a release from January this year. Can't vouch for it beyond that, so exploring other options later on may be desired.

The [Django docs recommend](https://docs.djangoproject.com/en/5.2/topics/i18n/translation/#gettext-on-windows) a different source of [binary installers](https://mlocati.github.io/articles/gettext-iconv-windows.html). I'm not sure if there's any reason to prefer one over the other.

>     * Changed the `llvm-cov` environment to use a temp shell script to get around shell piping issues on macOS using the `<( ... )` syntax.

Definitely out of scope for this PR, but I've been pondering the possibility of adding a [Justfile](https://just.systems/man/en/) with a way to run coverage locally. I tried it once but never got it working because of issues with the environment variables.

>     * Ignored one test on Windows because it looked like a can of worms I didn't necessarily want to open right now. 😄

Fair! I think this would be in scope for #27.

File: tests/test_django.py [hunk 3]
```diff
@@ -20,15 +21,16 @@ def test_parse_error():
         get_template("parse_error.txt", using="rusty")
 
     template_dir = Path("tests/templates").absolute()
+    path_separator = '/' if os.name != 'nt' else '\\'
```
Comment (LilyFoote) [inline tests/test_django.py:24]: We can use [`os.sep`](https://docs.python.org/3/library/os.html#os.sep) instead:

```suggestion
```

File: tests/test_django.py [hunk 4]
```diff
@@ -20,15 +21,16 @@ def test_parse_error():
...
   × Empty variable tag
-   ╭─[%s/parse_error.txt:1:28]
+   ╭─[%s%sparse_error.txt:1:28]
  1 │ This is an empty variable: {{ }}
    ·                            ──┬──
    ·                              ╰── here
    ╰────
 """
-    assert str(excinfo.value) == expected % template_dir
+    assert str(excinfo.value) == expected % (template_dir, path_separator)
```
Comment (LilyFoote) [inline tests/test_django.py:33]:
```suggestion
    assert str(excinfo.value) == expected % (template_dir, os.sep)
```

**Review by LilyFoote — APPROVED** [review]
Amazing work, thanks!

I see one place we can tidy up a bit and I see you've noticed a few others yourself, but I'm happy to land this when you are.

File: tests/test_django.py (hunk 3 above)
Comment (joshuadavidthomas) [reply tests/test_django.py:24]: TIL! They have everything in the stdlib, what will they think of next?!

**Comment by joshuadavidthomas** [comment]
> The [Django docs recommend](https://docs.djangoproject.com/en/5.2/topics/i18n/translation/#gettext-on-windows) a different source of [binary installers](https://mlocati.github.io/articles/gettext-iconv-windows.html). I'm not sure if there's any reason to prefer one over the other.

Oh, I just landed on the one I did because it was the first search result for "gettext Windows GitHub Actions". Differences are probably negligible, but I would lean towards using what's recommended by Django if only because it's in the docs for a reason. (Whether I know that reason or not is irrelevant to me. 😄) Should be a simple swap of the URLs in that step.

> Definitely out of scope for this PR, but I've been pondering the possibility of adding a [Justfile](https://just.systems/man/en/) with a way to run coverage locally. I tried it once but never got it working because of issues with the environment variables.

OH yes, I'm a big fan of just and [use] [Justfiles] [in] [all]() [of]() [my]() [projects](). I've tried a bunch of the patterns of how to layout Justfiles/use just and have liked where I've landed recently with modules to group commands. Even just putting one line simple commands behind a just commands helps improve the dev experience, since you can swap out the underlying command while keeping the CLI interface consistent. (E.g., swapping from pip-tools to uv for dependency management.) Big +1 from me.

Also interested in getting coverage on [my Django/Rust project], but I'm a relatively new to the Rust ecosystem so I haven't had the time to dig in to how to actually go about setting it up. I'm definitely going to be ~stealing~ borrowing some ideas 👀.

[use]: https://github.com/joshuadavidthomas/django-bird-autoconf/blob/main/Justfile
[Justfiles]: https://github.com/joshuadavidthomas/django-github-app/blob/main/Justfile
[in]: https://github.com/joshuadavidthomas/django-bird/blob/main/Justfile
[all]: https://github.com/joshuadavidthomas/django-language-server/blob/main/Justfile
[of]: https://github.com/joshuadavidthomas/django-lazy-gdal/blob/main/Justfile
[my]: https://github.com/joshuadavidthomas/llm-uv-tool/blob/main/Justfile
[projects]: https://github.com/westerveltco/django-twc-project/blob/main/examples/default/Justfile
[my Django/Rust project]: https://github.com/joshuadavidthomas/django-language-server
